app/src/main/java/com/willowtree/vocable/presets/adapter/AddToCategoryPickerViewModel.kt

```diff
@@ -0,0 +1,154 @@
+            if (isChecked) {
+                // add it to the category
+                addNewPhrase(phraseString, category)
+            } else { // if they un-checked the switch
+                // delete it from the category
+                deletePhraseFromCategory(phraseString, category)
+            }
+        }
+    }
+
+    fun buildCategoryList(phraseString: String) {
+        // map category to whether or not the phrase is in that category
+        val phraseMap = mutableMapOf<Category, Boolean>()
+
+        backgroundScope.launch {
+            userCategories = presetsRepository.getUserGeneratedCategories()
+
+            // for each category in the list
+            userCategories.forEach { category ->
+                // get all the phrases associated with it
+                val phrases = presetsRepository.getPhrasesForCategory(category.categoryId)
+
+                // for each phrase in the category
+                phrases.forEach {
+                    // if we find a match
+                    if (localizedResourceUtility.getTextFromPhrase(it) == phraseString) {
+                        // add it to the map
+                        phraseMap[category] = true
+                    }
+                }
+
+                // if we get through the loop without adding anything, the phrase isn't in there
+                if (phraseMap[category] == null) {
+                    phraseMap[category] = false
+                }
+            }
+
+            liveCategoryMap.postValue(phraseMap)
+        }
+    }
+
+    private fun deletePhraseFromCategory(phraseString: String, category: Category) {
+        backgroundScope.launch {
+            var phrase: Phrase? = null
+
+            val phrasesInCategory = presetsRepository.getPhrasesForCategory(category.categoryId)
+            for (p in phrasesInCategory) {
+                if (localizedResourceUtility.getTextFromPhrase(p) == phraseString) {
+                    phrase = p
+                }
+            }
+
+            with(presetsRepository) {
+                if (phrase != null) {
+                    val crossRefs = getCrossRefsForPhraseIds(listOf(phrase.phraseId))
+
+                    // if the phrase is in only one category, delete it from the db entirely
+                    if (crossRefs.size == 1) {
+                        deletePhrase(phrase)
+                    } else { // otherwise, just delete the one cross reference
+                        crossRefs.forEach { ref ->
+                            if (ref.categoryId == category.categoryId) {
+                                deleteCrossRef(ref)
+                            }
+                        }
+                    }
+
+                    // remove the category from the saved phrases map
+                    savedPhrases.remove(category.categoryId)
+                }
+            }
+
+            liveShowPhraseDeleted.postValue(true)
+            delay(KeyboardViewModel.PHRASE_ADDED_DELAY)
```

Comment:
I think this is an example of me not having enuff context on the project. Why we are intentionally adding a `delay` here? :o

Reply:
This is for a dialog popup - post a value of true to show the dialog, wait two seconds, then post a value of false to dismiss it.  Not something a decision I made, but it's used in a couple places.